Source/Core/Core/WatchDog.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/WatchDog.h"
+
+#include "Common/Assert.h"
+#include "Common/ChunkFile.h"
+#include "Common/Logging/Log.h"
+#include "Common/SPSCQueue.h"
+
+#include "VideoCommon/VideoConfig.h"
+
+#include <algorithm>
+#include <cmath>
+
+WatchDog::Profiler g_watchdog;
+WatchDog::Profiler g_gpu_watchdog;
+
+void WatchDog::Profiler::Reset(const Marker& first_marker)
+{
+  std::lock_guard lock{m_profiler_lock};
+
+  m_dt_total = DT::zero();
+  m_dt_queue.clear();
+
+  m_dt_marked_totals = {};
+
+  m_last_time = Clock::now();
+  m_last_marker = first_marker;
+}
+
+void WatchDog::Profiler::SetMarker(const Marker& next_marker)
+{
+  if (!g_ActiveConfig.bEnableProfiler)
+    return;
+
+  std::lock_guard lock{m_profiler_lock};
+
+  const TimePoint now = Clock::now();
+  const DT duration = now - m_last_time;
+  m_last_time = now;
+
+  const DT window =
+      std::chrono::duration_cast<DT>(DT_us(std::max(1, g_ActiveConfig.iPerfSampleUSec)));
+
+  if (window < duration)
+    return Reset(next_marker);
+
+  const TimeSlice slice{m_last_marker, duration};
+  m_dt_queue.push_back(slice);
+  m_dt_total += slice.duration;
+  m_dt_marked_totals[slice.marker] += slice.duration;
+
+  while (!m_dt_queue.empty() && window < m_dt_total)
+  {
+    const TimeSlice back = m_dt_queue.front();
+    m_dt_total -= back.duration;
+    m_dt_marked_totals.at(back.marker) -= back.duration;
+    m_dt_queue.pop_front();
+  }
+
+  m_last_marker = next_marker;
+}
+
+void WatchDog::Profiler::PushMarker(const Marker& next_marker)
+{
+  if (!g_ActiveConfig.bEnableProfiler)
+    return;
+
+  std::lock_guard lock{m_profiler_lock};
+
+  m_marker_stack.push_back(m_last_marker);
+  SetMarker(next_marker);
+}
+
+void WatchDog::Profiler::PopMarker()
+{
+  if (!g_ActiveConfig.bEnableProfiler)
+    return;
+
+  std::lock_guard lock{m_profiler_lock};
+
+  if (!m_marker_stack.empty())
+  {
+    SetMarker(m_marker_stack.back());
+    m_marker_stack.pop_back();
+  }
+}
+
+std::vector<WatchDog::ProfilerEntry> WatchDog::Profiler::GetEntries(std::size_t max_entries)
+{
+  if (!g_ActiveConfig.bEnableProfiler)
+    return {};
+
+  std::lock_guard lock{m_profiler_lock};
+  ScopedMarker sorting(*this, "Profiler Sorting");
+
+  static std::vector<WatchDog::ProfilerEntry> entries;
```

Comment:
I wonder if you gain anything by marking it static and then clearing right after as opposed to a new value..

Reply:
I can't really imagine it being that substantial, and if it does matter that the buffer isn't being reallocated that much, it's much more appropriate to make it a class member than marking it as static inside the function.